### PR TITLE
Kokkos new api unify

### PR DESCRIPTION
**Pull request by LuisAlfredoNu**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [X] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [X] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [X] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Following the design of **Lightning Qubit* uses the New Device API end to end, we can now migrate lightning.kokkos device to the new device API.

**Description of the Change:**
Increase the reusability of the code by creating base classes for Lightning StateVector, Measurements, and AjointJacobian. Create a new device base class for LightningQubit and LightningKokkos.

**Benefits:**
Improve the maintenance and readability of the Python front-end devices for LQ and LK.

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-71783]

### CODE REVIEW

File: pennylane_lightning/core/_adjoint_jacobian_base.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+from pennylane_lightning.core.lightning_base import _chunk_iterable
+
+
+class LightningBaseAdjointJacobian:
+    """Check and execute the adjoint Jacobian differentiation method.
+
+    Args:
+        qubit_state(Lightning[Device]StateVector): State Vector to calculate the adjoint Jacobian with.
+        batch_obs(bool): If serialized tape is to be batched or not.
+    """
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/core/_adjoint_jacobian_base.py:38]: AdjointJacobian base class. 

The classes and devices names, mentioned in base clasess doctrings, were replaced by `Lightning[Device]StateVector` or `Lightning[Device]AdjointJacobian`, etc.

File: pennylane_lightning/core/_adjoint_jacobian_base.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+    def __init__(self, qubit_state: Any, batch_obs: bool = False) -> None:
+        self._qubit_state = qubit_state
+        self._state = qubit_state.state_vector
+        self._dtype = qubit_state.dtype
+        self._batch_obs = batch_obs
+
+        # Dummy for the C++ bindings
+        self._jacobian_lightning: Callable = None
+        self._create_ops_list_lightning: Callable = None
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/core/_adjoint_jacobian_base.py:48]: Adding comments for dummy variables to help further implementation

File: pennylane_lightning/core/_measurements_base.py [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+    def dtype(self):
+        """Returns the simulation data type."""
+        return self._dtype
+
+    def _measurement_dtype(self):
+        """Binding to Lightning Kokkos Measurements C++ class.
+
+        Returns: the Measurements class
+        """
+        pass
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/core/_measurements_base.py:76]: Mock the method to set the C++ binding for each device.

File: pennylane_lightning/core/_state_vector_base.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        [0.+0.j 1.+0.j]
+        """
+        pass
+
+    def _state_dtype(self):
+        """Binding to Lightning Managed state vector C++ class.
+
+        Returns: the state vector class
+        """
+        pass
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/core/_state_vector_base.py:94]: Mock the class methods that should be implemented in all devices.

File: pennylane_lightning/core/lightning_newAPI_base.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+
+        if isinstance(wires, int):
+            self._wire_map = None  # should just use wires as is
+        else:
+            self._wire_map = {w: i for i, w in enumerate(self.wires)}
+
+        # Dummy for LightningStateVector, LightningMeasurements, LightningAdjointJacobian
+        self.LightningStateVector: Callable = None
+        self.LightningMeasurements: Callable = None
+        self.LightningAdjointJacobian: Callable = None
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/core/lightning_newAPI_base.py:84]: Dummy variables for the different classes needed for the Device

File: pennylane_lightning/core/lightning_newAPI_base.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+    def c_dtype(self):
+        """State vector complex data type."""
+        return self._c_dtype
+
+    dtype = c_dtype
+
+    def _set_Lightning_classes(self):
+        """Load the LightningStateVector, LightningMeasurements, LightningAdjointJacobian as class attribute"""
+        pass
+
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/core/lightning_newAPI_base.py:96]: Class method to define the other Device classes

File: pennylane_lightning/core/lightning_newAPI_base.py [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+    def simulate(
+        self,
+        circuit: QuantumScript,
+        state,  # LightningStateVector
+        postselect_mode: str = None,
+    ) -> Result:
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/core/lightning_newAPI_base.py:102]: The methods `simulate`, `jacobian`, vjp, `simulate_and_jacobian`, and `simulate_and_vjp` were moved inside the LightningBase class to increase the reusability.

File: pennylane_lightning/lightning_kokkos/_adjoint_jacobian.py [hunk 8]
```diff
@@ -25,315 +25,36 @@
...
-# pylint: disable=import-error, no-name-in-module, ungrouped-imports
-from pennylane_lightning.core._serialize import QuantumScriptSerializer
-from pennylane_lightning.core.lightning_base import _chunk_iterable
+from pennylane_lightning.core._adjoint_jacobian_base import LightningBaseAdjointJacobian
 
 from ._state_vector import LightningKokkosStateVector
 
 
-class LightningKokkosAdjointJacobian:
+class LightningKokkosAdjointJacobian(LightningBaseAdjointJacobian):
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/lightning_kokkos/_adjoint_jacobian.py:35]: The `LightningAdjointJacobian` classes were the easier to unify. :smile:

File: pennylane_lightning/lightning_kokkos/_state_vector.py [hunk 9]
```diff
@@ -57,33 +57,30 @@ class LightningKokkosStateVector:  # pylint: disable=too-few-public-methods
...
+        self._kokkos_config = {}
+        self._sync = sync
+
+        # Initialize the state vector
         if kokkos_args is None:
-            self._kokkos_state = self._state_dtype()(self.num_wires)
+            self._qubit_state = self._state_dtype()(self.num_wires)
         elif isinstance(kokkos_args, InitializationSettings):
-            self._kokkos_state = self._state_dtype()(self.num_wires, kokkos_args)
+            self._qubit_state = self._state_dtype()(self.num_wires, kokkos_args)
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/lightning_kokkos/_state_vector.py:83]: Rename the state variable to a generic name to be used across the other base classes

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py [hunk 10]
```diff
@@ -40,188 +38,34 @@
...
-    Note that this function can return measurements for non-commuting observables simultaneously.
-    """
-    if wire_map is not None:
-        [circuit], _ = qml.map_wires(circuit, wire_map)
-    res = simulate(circuit, state)
-    _vjp = LightningKokkosAdjointJacobian(state, batch_obs=batch_obs).calculate_vjp(
-        circuit, cotangents
-    )
-    return res, _vjp
-
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/lightning_kokkos/lightning_kokkos.py:223]: All these methods were moved into the `LightningBase` class to improve reusability.

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py [hunk 11]
```diff
@@ -467,24 +308,25 @@ def __init__(  # pylint: disable=too-many-arguments
...
+        # Set the attributes to call the Lightning classes
+        self._set_Lightning_classes()
 
         # Kokkos specific options
         self._kokkos_args = kokkos_args
         self._sync = sync
 
-        self._statevector = LightningKokkosStateVector(
+        # Creating the state vector
+        self._statevector = self.LightningStateVector(
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/lightning_kokkos/lightning_kokkos.py:329]: Using the defined attribute from `self._set_Lightning_classes()`

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py [hunk 12]
```diff
@@ -496,12 +338,11 @@ def name(self):
...
-    def c_dtype(self):
-        """State vector complex data type."""
-        return self._c_dtype
-
-    dtype = c_dtype
+    def _set_Lightning_classes(self):
+        """Load the LightningStateVector, LightningMeasurements, LightningAdjointJacobian as class attribute"""
+        self.LightningStateVector = LightningKokkosStateVector
+        self.LightningMeasurements = LightningKokkosMeasurements
+        self.LightningAdjointJacobian = LightningKokkosAdjointJacobian
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/lightning_kokkos/lightning_kokkos.py:345]: Passing the Lightning classes to attributes to be used in `LightningBase`

File: tests/lightning_qubit/test_jacobian_method.py [hunk 13]
```diff
@@ -16,28 +16,10 @@
...
-    )
-
-if device_name == "lightning.kokkos":
-    from pennylane_lightning.lightning_kokkos.lightning_kokkos import (
-        jacobian,
-        simulate,
-        simulate_and_jacobian,
-        simulate_and_vjp,
-        vjp,
-    )
```
Comment (LuisAlfredoNu) [inline tests/lightning_qubit/test_jacobian_method.py:39]: Now these methods are inside of `LightningDevice`

File: tests/lightning_qubit/test_simulate_method.py [hunk 14]
```diff
@@ -55,6 +46,11 @@ def calculate_reference(tape):
         results = dev.execute(tapes)
         return transf_fn(results)
 
+    @staticmethod
+    def calculate_result(wires, tape, statevector):
+        dev = LightningDevice(wires)
+        return dev.simulate(circuit=tape, state=statevector)
```
Comment (LuisAlfredoNu) [inline tests/lightning_qubit/test_simulate_method.py:52]: It is necessary to define the Device to get access to the `simulate` method

File: pennylane_lightning/core/lightning_newAPI_base.py [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+    # General device options
+    _new_API = True
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        wires,
+        *,
+        c_dtype=np.complex128,
+        shots=None,
+        batch_obs=False,
```
Comment (maliasadi) [inline pennylane_lightning/core/lightning_newAPI_base.py:69]: Sure, we can keep these common args but devices can define their own custom init args and kwargs..

File: pennylane_lightning/core/lightning_newAPI_base.py (hunk 5 above)
Comment (maliasadi) [inline pennylane_lightning/core/lightning_newAPI_base.py:84]: The base class doesn't need to know about these callables. Could we remove them from the base class?

File: pennylane_lightning/core/lightning_newAPI_base.py [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+        self.LightningAdjointJacobian: Callable = None
+
+    @property
+    def c_dtype(self):
+        """State vector complex data type."""
+        return self._c_dtype
+
+    dtype = c_dtype
+
+    def _set_Lightning_classes(self):
```
Comment (maliasadi) [inline pennylane_lightning/core/lightning_newAPI_base.py:93]:
```suggestion
    @abstractmethod
    def _set_Lightning_classes(self):
```

File: pennylane_lightning/core/_adjoint_jacobian_base.py [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+    def state(self):
+        """Returns a handle to the Lightning internal data object."""
+        return self._state
+
+    @property
+    def dtype(self):
+        """Returns the simulation data type."""
+        return self._dtype
+
+    def _adjoint_jacobian_dtype(self):
```
Comment (maliasadi) [inline pennylane_lightning/core/_adjoint_jacobian_base.py:65]: https://docs.python.org/3/library/abc.html#abc.abstractmethod

```suggestion
    @abstractmethod
    def _adjoint_jacobian_dtype(self):
```

**Review by maliasadi — CHANGES_REQUESTED** [review]
I left a few questions and suggestions, mostly related to the design of base classes, in my first pass. Happy to discuss them and finalize the work... Nice work! :100:

File: pennylane_lightning/core/lightning_newAPI_base.py
[the remark this answers is not in the document]
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/lightning_newAPI_base.py:1]: Yes, I was considering renaming this file as you suggested but already exist this file for the old device API. It might be better to wait until all the devices have been  migrated to the new device API

File: pennylane_lightning/core/_adjoint_jacobian_base.py [hunk 18]
```diff
@@ -0,0 +1,331 @@
...
+
+    Args:
+        qubit_state(Lightning[Device]StateVector): State Vector to calculate the adjoint Jacobian with.
+        batch_obs(bool): If serialized tape is to be batched or not.
+    """
+
+    def __init__(self, qubit_state: Any, batch_obs: bool = False) -> None:
+        self._qubit_state = qubit_state
+        self._state = qubit_state.state_vector
+        self._dtype = qubit_state.dtype
```
[the remark this answers is not in the document]
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/_adjoint_jacobian_base.py:43]: Yes, you are right. I will keep only `_qubit_state`

[the remark this answers is not in the document]
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/_adjoint_jacobian_base.py:43]: Also I will change the other files :rocket:

File: pennylane_lightning/core/_state_vector_base.py [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+        if not set(state.tolist()).issubset({0, 1}):
+            raise ValueError("BasisState parameter must consist of 0 or 1 integers.")
+
+        if len(state) != len(wires):
+            raise ValueError("BasisState parameter and wires must be of equal length.")
+
+        # Return a computational basis state over all wires.
+        self._qubit_state.setBasisState(list(state), list(wires))
+
+    def _apply_lightning_controlled(self, operation):
```
[the remark this answers is not in the document]
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/_state_vector_base.py:130]: Thank you for the suggestion, I did not know about these utilities. :rocket:

File: pennylane_lightning/core/lightning_newAPI_base.py (hunk 15 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/lightning_newAPI_base.py:65]: Yes, I add specific arg for each device but these ones are essential in all devices. Additionally, I will typing correctly each arg   :sweat_smile:

File: pennylane_lightning/core/lightning_newAPI_base.py (hunk 6 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/lightning_newAPI_base.py:92]: I think it will not be necessary `@abstractclassmethod` in this case but it is good to know about it. Thank you very much

File: pennylane_lightning/core/lightning_newAPI_base.py [hunk 20]
```diff
@@ -0,0 +1,348 @@
...
+
+    def jacobian(self, circuit: QuantumTape, state, batch_obs=False, wire_map=None):
+        """Compute the Jacobian for a single quantum script.
+
+        Args:
+            circuit (QuantumTape): The single circuit to simulate
+            state (LightningKokkosStateVector): handle to the Lightning state vector
+            batch_obs (bool): Determine whether we process observables in parallel when
+                computing the jacobian. This value is only relevant when the lightning
+                kokkos is built with OpenMP. Default is False.
```
[the remark this answers is not in the document]
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/lightning_newAPI_base.py:123]: My apologies, I will double check all the docstrings again.

File: pennylane_lightning/core/lightning_newAPI_base.py (hunk 5 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/lightning_newAPI_base.py:80]: Why do you suggest removing it? :thinking:

File: pennylane_lightning/core/lightning_newAPI_base.py [hunk 21]
```diff
@@ -0,0 +1,348 @@
...
+    Args:
+        wires (int): the number of wires to initialize the device with
+        sync (bool): immediately sync with host-sv after applying operations
+        c_dtype: Datatypes for statevector representation. Must be one of
+            ``np.complex64`` or ``np.complex128``.
+        shots (int): How many times the circuit should be evaluated (or sampled) to estimate
+            the expectation values. Defaults to ``None`` if not specified. Setting
+            to ``None`` results in computing statistics like expectation values and
+            variances analytically.
+        kokkos_args (InitializationSettings): binding for Kokkos::InitializationSettings
```
[the remark this answers is not in the document]
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/lightning_newAPI_base.py]: I will remove the comment from the base class related to a specific device :+1:

File: pennylane_lightning/core/lightning_newAPI_base.py [hunk 22]
```diff
@@ -35,7 +35,7 @@
 
 @simulator_tracking
 @single_tape_support
-class LightningBase(ABC, Device):
```
Comment (AmintorDusko) [inline pennylane_lightning/core/lightning_newAPI_base.py]: If inheritance order matters we probably have methods with the same name in both parents. Is this the case? Do you happen to know which?

Comment (LuisAlfredoNu) [reply pennylane_lightning/core/lightning_newAPI_base.py]: I checked and `Device` has the parent `ABC` too. In this case, I will remove the parent class `ABC` from the `LightningBase`. Thank you @AmintorDusko for the catch :rocket:

File: pennylane_lightning/core/lightning_newAPI_base.py (hunk 5 above)
Comment (maliasadi) [reply pennylane_lightning/core/lightning_newAPI_base.py:80]: I'd like to generalize the base class as much as possible. I wonder if these dummy callables can be abstracted away in anyways. But, I think we need them right now to preserve consistency between the method signature in base and concrete device classes.

**Comment by AmintorDusko** [comment]
@LuisAlfredoNu, it looks like you still have problems with the formatting check.

**Review by maliasadi — APPROVED** [review]
Great work @LuisAlfredoNu! :100:

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py [hunk 23]
```diff
@@ -40,188 +38,34 @@
...
-    res = simulate(circuit, state)
-    _vjp = LightningKokkosAdjointJacobian(state, batch_obs=batch_obs).calculate_vjp(
-        circuit, cotangents
-    )
-    return res, _vjp
-
+# Result_or_ResultBatch = Union[Result, ResultBatch]
+# QuantumTapeBatch = Sequence[QuantumTape]
+# QuantumTape_or_Batch = Union[QuantumTape, QuantumTapeBatch]
+# PostprocessingFn = Callable[[ResultBatch], Result_or_ResultBatch]
```
Comment (AmintorDusko) [inline pennylane_lightning/lightning_kokkos/lightning_kokkos.py:66]: Are we leaving this here?

File: pennylane_lightning/lightning_qubit/_state_vector.py [hunk 24]
```diff
@@ -49,43 +52,15 @@ class LightningStateVector:
     """
 
     def __init__(self, num_wires, dtype=np.complex128, device_name="lightning.qubit"):
```
Comment (AmintorDusko) [inline pennylane_lightning/lightning_qubit/_state_vector.py:54]: We don't need the device name as an input variable anymore. The original idea was to have the device_name as the agent responsible for the multi-dispatching. As we move away from this approach, the device name can be an internal attribute, if needed.

File: pennylane_lightning/lightning_qubit/_state_vector.py [hunk 25]
```diff
@@ -49,43 +52,15 @@ class LightningStateVector:
...
-        self._num_wires = num_wires
-        self._wires = Wires(range(num_wires))
-        self._dtype = dtype
-
-        if dtype not in [np.complex64, np.complex128]:  # pragma: no cover
-            raise TypeError(f"Unsupported complex type: {dtype}")
-
         if device_name != "lightning.qubit":
             raise DeviceError(f'The device name "{device_name}" is not a valid option.')
```
Comment (AmintorDusko) [inline pennylane_lightning/lightning_qubit/_state_vector.py:57]: Similar to device_name, this is no longer necessary.

File: pennylane_lightning/lightning_qubit/_state_vector.py [hunk 26]
```diff
@@ -49,43 +52,15 @@ class LightningStateVector:
...
-    def wires(self):
-        """All wires that can be addressed on this device"""
-        return self._wires
+        super().__init__(num_wires, dtype)
 
-    @property
-    def num_wires(self):
-        """Number of wires addressed on this device"""
-        return self._num_wires
+        self._device_name = device_name
```
Comment (AmintorDusko) [inline pennylane_lightning/lightning_qubit/_state_vector.py:60]: Please check if this parameter is still useful.

File: pennylane_lightning/lightning_qubit/lightning_qubit.py [hunk 27]
```diff
@@ -37,188 +37,33 @@
...
-    if wire_map is not None:
-        [circuit], _ = qml.map_wires(circuit, wire_map)
-    res = simulate(circuit, state)
-    _vjp = LightningAdjointJacobian(state, batch_obs=batch_obs).calculate_vjp(circuit, cotangents)
-    return res, _vjp
-
+# Result_or_ResultBatch = Union[Result, ResultBatch]
+# QuantumTapeBatch = Sequence[QuantumTape]
+# QuantumTape_or_Batch = Union[QuantumTape, QuantumTapeBatch]
+# PostprocessingFn = Callable[[ResultBatch], Result_or_ResultBatch]
```
Comment (AmintorDusko) [inline pennylane_lightning/lightning_qubit/lightning_qubit.py:64]: Do we need that?

File: tests/lightning_qubit/test_jacobian_method.py [hunk 28]
```diff
@@ -68,13 +50,14 @@ def calculate_reference(tape, execute_and_derivatives=False):
         return transf_fn(results), jac
 
     @staticmethod
-    def process_and_execute(statevector, tape, execute_and_derivatives=False):
+    def process_and_execute(wires, statevector, tape, execute_and_derivatives=False):
```
Comment (AmintorDusko) [inline tests/lightning_qubit/test_jacobian_method.py:53]: You don't need to add `wires` as an input here, as you can query this information directly from the statevector, right?

File: tests/lightning_qubit/test_jacobian_method.py [hunk 29]
```diff
@@ -144,14 +127,15 @@ def calculate_reference(tape, dy, execute_and_derivatives=False):
         return transf_fn(results), jac
 
     @staticmethod
-    def process_and_execute(statevector, tape, dy, execute_and_derivatives=False):
+    def process_and_execute(wires, statevector, tape, dy, execute_and_derivatives=False):
```
Comment (AmintorDusko) [inline tests/lightning_qubit/test_jacobian_method.py:130]: Same here. the number of wires can be directly queried from the statevector class.

File: tests/lightning_qubit/test_simulate_method.py [hunk 30]
```diff
@@ -55,6 +46,11 @@ def calculate_reference(tape):
         results = dev.execute(tapes)
         return transf_fn(results)
 
+    @staticmethod
+    def calculate_result(wires, tape, statevector):
```
Comment (AmintorDusko) [inline tests/lightning_qubit/test_simulate_method.py:50]: Similarly as before the number of wires can be directly taken from the statevector.

**Review by AmintorDusko — CHANGES_REQUESTED** [review]
Looks pretty good! It needs a few adjustments, though. Could you re-tag me when done and I will give you one more look? Thanks!

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py (hunk 23 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/lightning_kokkos/lightning_kokkos.py:66]: I will remove it

File: pennylane_lightning/lightning_qubit/_state_vector.py (hunk 24 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/lightning_qubit/_state_vector.py:54]: I will remove and change this.

File: pennylane_lightning/core/_adjoint_jacobian_base.py [hunk 31]
```diff
@@ -0,0 +1,290 @@
...
+    Check and execute the adjoint Jacobian differentiation method.
+
+    Args:
+        qubit_state(Lightning[Device]StateVector): State Vector to calculate the adjoint Jacobian with.
+        batch_obs(bool): If serialized tape is to be batched or not.
+    """
+
+    def __init__(self, qubit_state: Any, batch_obs: bool) -> None:
+        self._qubit_state = qubit_state
+        self._batch_obs = batch_obs
```
Comment (maliasadi) [inline pennylane_lightning/core/_adjoint_jacobian_base.py:44]: @LuisAlfredoNu Could you add tests for the base class methods?

Comment (maliasadi) [reply pennylane_lightning/core/_adjoint_jacobian_base.py:44]: We should have unit tests for all these methods. They don't need to be complicated, but ensure that you test the expected logic and corner cases.

File: pennylane_lightning/core/_adjoint_jacobian_base.py [hunk 32]
```diff
@@ -0,0 +1,290 @@
...
+
+    def _process_jacobian_tape(
+        self, tape: QuantumTape, use_mpi: bool = False, split_obs: bool = False
+    ):
+        """Process a tape, serializing and building a dictionary proper for
+        the adjoint Jacobian calculation in the C++ layer.
+
+        Args:
+            tape (QuantumTape): Operations and measurements that represent instructions for execution on Lightning.
+            use_mpi (bool, optional): If using MPI to accelerate calculation. Defaults to False.
```
Comment (maliasadi) [inline pennylane_lightning/core/_adjoint_jacobian_base.py:100]: Should this be here? I think this should be removed too:
```suggestion
```

File: pennylane_lightning/core/_adjoint_jacobian_base.py [hunk 33]
```diff
@@ -0,0 +1,290 @@
...
+        if not measurements:
+            return None
+
+        if len(measurements) == 1 and measurements[0].return_type is State:
+            return State
+
+        return Expectation
+
+    def _process_jacobian_tape(
+        self, tape: QuantumTape, use_mpi: bool = False, split_obs: bool = False
```
Comment (maliasadi) [inline pennylane_lightning/core/_adjoint_jacobian_base.py:93]: Not all devices have support for mpi.. this should be optional.
```suggestion
        self, tape: QuantumTape, split_obs: bool = False
```

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py [hunk 34]
```diff
@@ -467,24 +308,25 @@ def __init__(  # pylint: disable=too-many-arguments
             raise ImportError(
                 "Pre-compiled binaries for lightning.kokkos are not available. "
                 "To manually compile from source, follow the instructions at "
-                "https://pennylane-lightning.readthedocs.io/en/latest/installation.html."
+                "https://docs.pennylane.ai/projects/lightning/en/stable/dev/installation.html."
```
Comment (vincentmr) [inline pennylane_lightning/lightning_kokkos/lightning_kokkos.py:311]: Goot catch.

File: pennylane_lightning/core/_adjoint_jacobian_base.py [hunk 35]
```diff
@@ -0,0 +1,290 @@
...
+            ):
+                # We now just ignore non-op or state preps
+                tp_shift.append(trainable_param)
+                record_tp_rows.append(all_params)
+            all_params += 1
+
+        if use_sp:
+            # When the first element of the tape is state preparation. Still, I am not sure
+            # whether there must be only one state preparation...
+            tp_shift = [i - 1 for i in tp_shift]
```
Comment (maliasadi) [inline pennylane_lightning/core/_adjoint_jacobian_base.py:142]: Do we need this logic for all devices though?

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py [hunk 36]
```diff
@@ -467,24 +308,25 @@ def __init__(  # pylint: disable=too-many-arguments
...
-
-        if isinstance(wires, int):
-            self._wire_map = None  # should just use wires as is
-        else:
-            self._wire_map = {w: i for i, w in enumerate(self.wires)}
+        super().__init__(
+            wires=wires,
+            c_dtype=c_dtype,
+            shots=shots,
+            batch_obs=batch_obs,
```
Comment (vincentmr) [inline pennylane_lightning/lightning_kokkos/lightning_kokkos.py:318]: Setting `batch_obs` to false might be safer since it isn't supported in LK.
```suggestion
            batch_obs=False,
```

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py [hunk 37]
```diff
@@ -467,24 +308,25 @@ def __init__(  # pylint: disable=too-many-arguments
...
+        )
 
-        self._c_dtype = c_dtype
-        self._batch_obs = batch_obs
+        # Set the attributes to call the Lightning classes
+        self._set_Lightning_classes()
```
Comment (vincentmr) [inline pennylane_lightning/lightning_kokkos/lightning_kokkos.py:322]: I would conform to snake case here, even if Lightning is a proper name.
```suggestion
        self._set_lightning_classes()
```

File: pennylane_lightning/core/_measurements_base.py [hunk 38]
```diff
@@ -0,0 +1,398 @@
...
+        return self._qubit_state
+
+    @property
+    def dtype(self):
+        """Returns the simulation data type."""
+        return self._qubit_state.dtype
+
+    @abstractmethod
+    def _measurement_dtype(self):
+        """Binding to Lightning [Device] Measurements C++ class.
```
Comment (maliasadi) [inline pennylane_lightning/core/_measurements_base.py:74]: You use both `Lightning [Device] Measurements` and others `Lightning[Device]Measurements`. let's update this to one version: `Lightning [Device] Measurements` the same for StateVector and Jacobian base classes.

File: pennylane_lightning/lightning_qubit/_state_vector.py (hunk 26 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/lightning_qubit/_state_vector.py:60]: I checked and It is no longer needed. :smile:

File: pennylane_lightning/lightning_qubit/lightning_qubit.py (hunk 27 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/lightning_qubit/lightning_qubit.py:64]: I will delete it

File: pennylane_lightning/core/_adjoint_jacobian_base.py
Comment (maliasadi) [inline pennylane_lightning/core/_adjoint_jacobian_base.py:1]: What about adding these base classes to `lightning_base` subdirectory?

File: tests/lightning_qubit/test_jacobian_method.py
Comment (maliasadi) [inline tests/lightning_qubit/test_jacobian_method.py:1]: We definitely need tests for base classes :)

File: pennylane_lightning/lightning_qubit/_adjoint_jacobian.py [hunk 39]
```diff
@@ -235,98 +108,19 @@ def calculate_jacobian(self, tape: QuantumTape):
             processed_data["ops_serialized"],
             trainable_params,
         )
+
         jac = np.array(jac)
-        has_shape0 = bool(len(jac))
+        has_shape0: bool = len(jac) > 0
```
Comment (vincentmr) [inline pennylane_lightning/lightning_qubit/_adjoint_jacobian.py:113]: We do not usually use type annotations outside function interfaces, but I don't mind if we start doing it.
```suggestion
        has_shape0 = len(jac) > 0
```

**Review by maliasadi — CHANGES_REQUESTED** [review]
I had a second pass over the code... Thanks @LuisAlfredoNu for addressing my first round of review comments by using `abc` & better practices for writing base 'abstract' classes in Python. However, I am afraid there are still a few more things to fix.

**Review by vincentmr — APPROVED** [review]
Nice work @LuisAlfredoNu , I love the code reduction! A few suggestions, but otherwise LGTM.

File: tests/lightning_qubit/test_jacobian_method.py (hunk 28 above)
Comment (LuisAlfredoNu) [reply tests/lightning_qubit/test_jacobian_method.py:53]: Yes, you are right. I will fix it

File: pennylane_lightning/lightning_qubit/_state_vector.py (hunk 26 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/lightning_qubit/_state_vector.py:60]: I double-checked. However, running the test suit I discovered that `self._device_name` is needed for `QuantumScriptSerializer`

File: pennylane_lightning/lightning_qubit/_adjoint_jacobian.py (hunk 39 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/lightning_qubit/_adjoint_jacobian.py:113]: Yes, I added it because I remember that It was confusing at some point but I think it is better to keep simple

File: tests/lightning_qubit/test_jacobian_method.py
Comment (maliasadi) [reply tests/lightning_qubit/test_jacobian_method.py:1]: For testing these base classes, you can create a dummy (mock) class derived from these base classes and test the expected behaviour for
- abstract methods that are not implemented in the dummy class (what would be the expected raised error messages?)
- dunder and helper methods with minimul logic
- abstract methods with a simple logic

If there are methods with complicated implementation, we should discuss:
- should we remove this from the base class or
- can the logic be simplified or fully used in all lightning devices?

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py [hunk 40]
```diff
@@ -467,24 +304,30 @@ def __init__(  # pylint: disable=too-many-arguments
...
                 "To manually compile from source, follow the instructions at "
-                "https://pennylane-lightning.readthedocs.io/en/latest/installation.html."
+                "https://docs.pennylane.ai/projects/lightning/en/stable/dev/installation.html."
             )
 
-        super().__init__(wires=wires, shots=shots)
+        if batch_obs:
+            raise DeviceError(
+                "Lightning Kokkos does not support batch observables option. 'batch_obs' should be False"
+            )
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/lightning_kokkos/lightning_kokkos.py:313]: @vincentmr What do you think about making explicit the limitation of `batch_obs` for LK? Also, I think this will help the user

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py (hunk 36 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/lightning_kokkos/lightning_kokkos.py:319]: I reverted this change because I got complaints from pylint. However,  `batch_obs` is defined as False in `init` of LK class

**Comment by LuisAlfredoNu** [comment]
@AmintorDusko  All the suggestions are done or commented on. Could you make a second review, please? :smile:

File: pennylane_lightning/core/_adjoint_jacobian_base.py (hunk 35 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/_adjoint_jacobian_base.py:141]: For LQ and LK is still necessary. If I remove it, the test fails.

File: pennylane_lightning/core/_adjoint_jacobian_base.py (hunk 31 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/_adjoint_jacobian_base.py:44]: As we discussed offline, the creation of a test for the base class will be postponed until the migration of LGPU to the new device API.

File: tests/lightning_qubit/test_jacobian_method.py
Comment (LuisAlfredoNu) [reply tests/lightning_qubit/test_jacobian_method.py:1]: As we discussed offline, the creation of a test for the base class will be postponed until the migration of LGPU to the new device API.

File: pennylane_lightning/lightning_kokkos/lightning_kokkos.py (hunk 40 above)
Comment (vincentmr) [reply pennylane_lightning/lightning_kokkos/lightning_kokkos.py:313]: I would not change the behaviour in this PR, but I don't mind either way. Let's see if anyone else had an opinion.

**Review by maliasadi — APPROVED** [review]
Thanks @LuisAlfredoNu :partying_face: Please wait for @ringo-but-quantum to autoupdate the version before merging the PR

**Review by AmintorDusko — APPROVED** [review]
Great job! Thank you for that!
